## Task 4.8.

In [3]:
### Import libraries 
import pandas as pd
import numpy as np
import os

In [4]:
### Set path
path=r'C:\Users\User\Achievement4\Data'

In [5]:
### Import dataframe ('ords_prods_merge_derived')
ords_prods_merge = pd.read_pickle(os.path.join(path, 'Prepared Data', 'orders_products_merged_derived.pkl'))

In [6]:
### finding the aggregated mean of the “order_number” column grouped by “department_id”
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

### Overall Trends: The full dataset typically exhibits higher average order numbers across departments, suggesting it reflects a wider spectrum of customer behavior compared to the subset.
Key Differences: Department 21 has the highest average in both datasets but is higher in the subset (25.53) than in the full dataset (22.90). Meanwhile, Department 10 shows a significant increase in the full dataset (20.19 compared to 18.68).

In [7]:
### Creating a loyalty flag for existing customers using the transform() and loc() functions
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

C:\Users\User\AppData\Local\Temp\ipykernel_11148\3148727557.py:2: FutureWarning: The provided callable <function max at 0x0000024C997BAC00> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


### Checking the Output
ords_prods_merge.head(15)

In [12]:
### Creating loyality flag for customers
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [13]:
### Checking output
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  hour_of_day  \
0   3139998      138            28                   6           11   
1   1977647      138            30                   6           17   
2    389851      709             2                   0           21   
3    652770      764             1                   3           13   
4   1813452      764             3                   4           17   

   days_since_prior_order  first_order  add_to_cart_order  reordered _merge  \
0                     3.0        False                  5          0   both   
1                    20.0        False                  1          1   both   
2                     6.0        False                 20          0   both   
3                     NaN         True                 10          0   both   
4                     9.0        False                 11          1   both   

  price_range_loc     busiest_day  max_order      loyalty_flag  
0             NaN  Regularly busy         32  Regular customer  
1             NaN  Regularly busy         32  Regular customer  
2             NaN     Busiest day          5      New customer  
3             NaN  Regularly busy          3      New customer  
4             NaN      Least busy          3      New customer

In [11]:
### checking the basic statistics of the product prices for each loyalty category
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean']})

prices
                       mean
loyalty_flag               
Loyal customer    10.386336
New customer      13.294670
Regular customer  12.495717

### loyal customer purchase low-price items, and the new customers, on the opposite, tend to buy more expensive items. Regular customers are in the middle of this list, spending about 12.5$ on an item 

## Creating Spending Flag

In [17]:
### Aggregating Data with transform() Function
ords_prods_merge['avg_price'] =ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

C:\Users\User\AppData\Local\Temp\ipykernel_11148\1411250628.py:2: FutureWarning: The provided callable <function mean at 0x0000024C997BB600> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['avg_price'] =ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [18]:
### checking the columns "user_id" and the new created "avg_price"
ords_prods_merge[['user_id', 'avg_price']].head(10)

user_id  avg_price
0      138   6.935811
1      138   6.935811
2      709   7.930208
3      764   4.972414
4      764   4.972414
5      777   6.935398
6      825   5.957576
7      910   6.680000
8     1052   7.162500
9     1052   7.162500

In [19]:
### Creating spending flag for customers
ords_prods_merge.loc[ords_prods_merge['avg_price'] >= 10, 'spending_flag'] = 'High_spender'
ords_prods_merge.loc[ords_prods_merge['avg_price'] < 10, 'spending_flag'] = 'Low_spender'

In [21]:
### checking the new created column
ords_prods_merge[['user_id', 'avg_price', 'spending_flag']].head(10)

user_id  avg_price spending_flag
0      138   6.935811   Low_spender
1      138   6.935811   Low_spender
2      709   7.930208   Low_spender
3      764   4.972414   Low_spender
4      764   4.972414   Low_spender
5      777   6.935398   Low_spender
6      825   5.957576   Low_spender
7      910   6.680000   Low_spender
8     1052   7.162500   Low_spender
9     1052   7.162500   Low_spender

In [23]:
# Examining Value Count on 'spending_flag' Column
ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low_spender     31770614
High_spender      634245
Name: count, dtype: int64

## Creating Order Frequency Flag

In [24]:
### Aggregating Data with transform() Function
ords_prods_merge['median_days'] =ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

C:\Users\User\AppData\Local\Temp\ipykernel_11148\620265464.py:2: FutureWarning: The provided callable <function median at 0x0000024C998F72E0> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['median_days'] =ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [25]:
### checking the columns "user_id" and the new created "median_days"
ords_prods_merge[['user_id', 'median_days']].head(10)

user_id  median_days
0      138          8.0
1      138          8.0
2      709          8.0
3      764          9.0
4      764          9.0
5      777         11.0
6      825         20.0
7      910          6.0
8     1052         10.0
9     1052         10.0

In [32]:
### Creating Frequency flag for customers
ords_prods_merge.loc[ords_prods_merge['median_days'] >20, 'frequency_flag'] = 'Non-frequent customer'
ords_prods_merge.loc[(ords_prods_merge['median_days'] <= 20) & (ords_prods_merge['median_days'] > 10), 'frequency_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['median_days'] <=10, 'frequency_flag'] = 'Frequent customer'

In [40]:
### Exporting Dataframe 
ords_prods_merge.to_pickle(os.path.join(path, 'Prepared Data', 'orders_products_merged_grouped.pkl'))

In [ ]:
### checking the new created column
